optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -15,180 +15,207 @@
-    use_flash_attention: Optional[bool] = False,
-    flash_attention_recompute: Optional[bool] = False,
-):
-    """
-    Copied from MptBlock.forward: https://github.com/huggingface/transformers/blob/v4.32.0/src/transformers/models/mpt/modeling_mpt.py
-    The only differences are:
-    - add new args token_idx
-    - add new args use_flash_attention
-    - add new arg flash_attention_recompute
-    """
-    # hidden_states: [batch_size, seq_length, hidden_size]
-    # Layer norm at the beginning of the transformer layer.
-    layernorm_output = self.norm_1(hidden_states)
-
-    residual = hidden_states
-
-    # Self attention.
-    attn_outputs, attn_weights, past_key_value = self.attn(
-        layernorm_output,
-        position_bias=position_bias,
-        attention_mask=attention_mask,
-        past_key_value=layer_past,
-        token_idx=token_idx,
-        use_flash_attention=use_flash_attention,
-        flash_attention_recompute=flash_attention_recompute,
-    )
-
-    hidden_states = self.resid_attn_dropout(attn_outputs) + residual
-
-    layernorm_output = self.norm_2(hidden_states)
-
-    # Get residual
-    residual = hidden_states
-
-    # MLP.
-    output = self.ffn(layernorm_output, residual)
-    outputs = (output,)
-
-    if use_cache:
-        outputs += (past_key_value,)
-
-    if output_attentions:
-        outputs += (attn_weights,)
-
-    return outputs  # hidden_states, present, attentions
+            if len(position_bias.shape) != 3:
+                raise ValueError(f"Expecting position_bias shape to be 3 dimensions, got {len(position_bias.shape)}")
+            key_length = key_states.shape[-2]
+
+            position_bias_query_index = max(0, position_bias.size(1) - query_length)
+            position_bias_key_index = max(0, position_bias.size(2) - key_length)
+
+            position_bias = position_bias[:, position_bias_query_index:, position_bias_key_index:]
+
+        if use_flash_attention and FusedSDPA:
+            import habana_frameworks.torch.hpu as ht
+
+            with ht.sdp_kernel(enable_recompute=flash_attention_recompute):
+                attn_output = FusedSDPA.apply(
+                    query_states,
+                    key_states,
+                    value_states,
+                    attention_mask * torch.finfo(query_states.dtype).min + position_bias.to(query_states.dtype),
+                    0.0,
+                    False,
+                    None,
+                )
+            attn_weights = None
+        else:
+            attention_scores = torch.matmul(query_states, key_states.transpose(-1, -2)) * self.softmax_scale
+
+            if position_bias is not None:
+                attention_scores = attention_scores + position_bias
+            if attention_mask is not None:
+                attention_scores = attention_scores.masked_fill(attention_mask, torch.finfo(query_states.dtype).min)
+
+            # (batch_size, n_heads, seq_length, key_length)
+            if self.is_fp8:
```

Comment:
I think this part should be taken care of internally to use fp8 or bf16 kernel based on the setting. Did you actually see the fp8 kernel is called during bf16 run without this code? 
 
Falcon example. 
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/falcon/modeling_falcon.py#L180

Reply:
I confirmed softmax_fp8 take care fp8 and bf16 cases. code is updated.